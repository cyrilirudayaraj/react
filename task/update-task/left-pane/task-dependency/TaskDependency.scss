@import '~@athena/forge/sass/forge-abstracts';
.task-dependency {
  .task-dependency-header {
    border-bottom: 1px;
    border-color: #e5e5e5;
    border-bottom-style: solid;
    display: flex;
    justify-content: space-between;
    .task-dependency-header-text {
      left: 54px;
      position: absolute;
    }
    .status {
      background-color: #808080 !important;
    }
    .status_nowrap_no_transform {
      white-space: nowrap;
      text-transform: none;
      width: 125px;
      height: 24px;
      text-align: center;
    }
  }
  .test-with {
    margin-top: 25px;
  }
  .blocked-by {
    margin-top: 25px;
  }
  .no-tasks {
    padding-top: 15px;
    text-align: center;
    margin-bottom: auto;
  }
  .add-dependency {
    padding-top: 6px;
    padding-right: 50px;
    text-align: right;
  }
  .dependency-item {
    border-bottom: 1px;
    border-color: #e5e5e5;
    border-bottom-style: solid;
    .dependency-row {
      padding-top: 12px;
      padding-bottom: 12px;
      .dependency-content {
        .fe_c_icon {
          height: 22px;
          width: 22px;
          margin-left: 5px;
          vertical-align: top;
        }
      }
      .info-icon {
        height: 18px;
        width: 18px;
      }
      .status {
        background-color: #808080 !important;
      }
      .status.completed {
        background-color: #118647 !important;
      }
      .status_nowrap_no_transform {
        white-space: nowrap;
        text-transform: none;
        width: 130px;
        height: 24px;
        text-align: center;
        font-size: 12px;
      }
      .deployment-date {
        font-size: 14px;
      }
      .edit-dependency {
        padding-left: 0px;
        max-width: 10%;
      }
      .edit-icon,
      .delete-icon {
        display: none;
        height: auto;
      }
    }
    .error {
      margin-left: map-get($global-margin, 'medium');
      margin-top: map-get($global-margin, 'small');
    }
    .dependency-row:hover .edit-icon,
    .edit-icon:hover {
      display: inline-block;
    }
    .dependency-row:hover {
      background-color: #d9ebfa;
    }
    .dependency-row:hover .delete-icon,
    .delete-icon:hover {
      display: inline-block;
    }
  }
  .hidden-row {
    display: none !important;
  }
  .task-dependency-details {
    padding-bottom: map-get($global-padding, 'large');
  }
  .fe_l_grid-row__col {
    padding-right: 0px;
  }
  .status-tag {
    text-align: right;
    padding-right: 10px;
  }
  .fe_c_status-tag {
    font-size: 12px !important;
    line-height: 1.2;
    padding: 4px 6px;
  }
}

.add-dependency,
.edit-dependency {
  .fe_l_grid-row__col {
    width: 100% !important;
    min-width: 100%;
    padding-right: 15px;
    .fe_c_form-field__input-slot {
      min-width: 100% !important;
      max-width: 100%;
      .fe_c_input,
      .fe_c_textarea {
        width: 100% !important;
      }
      .fe_c_textarea {
        height: 88px !important;
      }
    }
    .fe_c_form-field__input-slot > .fe_c_select {
      width: 100% !important;
    }
  }

  .fe_c_form-field.row2 {
    width: 25%;
  }

  .fe_c_form-field.notes {
    width: 40%;
  }

  .root-add-button {
    padding-top: 0px;
  }

  .add-button,
  .delete-button {
    margin-right: 10px;
    margin-top: 28px;
    box-shadow: none !important;
  }

  .row {
    display: flex;
    align-items: flex-start;
  }
  .dependency-task {
    border-bottom: 1px solid rgb(223, 223, 223);
  }
  .pointer {
    cursor: pointer;
  }
}
