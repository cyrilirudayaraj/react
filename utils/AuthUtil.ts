const winObj: any = window;

class AuthUtil {
  static getLoggedInUsername(): string {
    return winObj.globalVars
      ? winObj.globalVars.username
      : process.env.REACT_APP_USER;
  }

  static isPermissionEnabled(): boolean {
    const enablePermission = winObj.globalVars
      ? winObj.globalVars.enablePermission
      : process.env.REACT_APP_ENABLE_PERMISSION;
    return enablePermission && enablePermission == 1;
  }

  static isAuthorized(
    permission: string,
    userPermissionList: string[]
  ): boolean {
    let result = true;
    const eventList = userPermissionList ? userPermissionList : [];
    if (AuthUtil.isPermissionEnabled()) {
      if (permission && !eventList.includes(permission)) {
        result = false;
      }
    }
    return result;
  }

  static isConsecutiveSignoffDisabled(): boolean {
    const disableConsecutiveSignoff = winObj.globalVars
      ? winObj.globalVars.disableConsecutiveSignoff
      : process.env.REACT_APP_DISABLE_CONSECUTIVE_SIGNOFF;
    return disableConsecutiveSignoff && disableConsecutiveSignoff == 1;
  }

  static disableConsecutiveSignoff(previousSignoffUsername: string): boolean {
    if (
      AuthUtil.isConsecutiveSignoffDisabled() &&
      AuthUtil.getLoggedInUsername() === previousSignoffUsername
    ) {
      return true;
    }
    return false;
  }
}

export default AuthUtil;
