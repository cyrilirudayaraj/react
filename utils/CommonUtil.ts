import moment from 'moment';
import AppConstants from '../constants/AppConstants';
import { TaskStep } from '../types';

class CommonUtil {
  static handleDateChange(event: any, formik: any): void {
    const { value } = event.target;
    let output: any = undefined;
    if (value) {
      const momentDate = moment(
        value,
        AppConstants.UI_CONSTANTS.DATE_INPUT_FORMAT,
        true
      );
      if (momentDate.isValid()) {
        output = momentDate.toDate();
      }
    }
    formik.setFieldValue(event.target.id, output, false);
  }
  static getPreviousTaskStep(taskSteps: any, activeTaskStep: any): any {
    const previousStep = taskSteps.filter(
      (taskStep: TaskStep) =>
        parseInt(taskStep.ordering) === parseInt(activeTaskStep.ordering) - 1
    );
    return previousStep[0];
  }

  static isActiveTaskType(taskTypeId: string | undefined): boolean {
    if (
      taskTypeId ===
        AppConstants.SERVER_CONSTANTS.TASK_TYPES.BUSINESS_REQUIREMENT_UPDATE ||
      taskTypeId ===
        AppConstants.SERVER_CONSTANTS.TASK_TYPES.NEW_BUSINESS_REQUIREMENT ||
      taskTypeId ===
        AppConstants.SERVER_CONSTANTS.TASK_TYPES.DUAL_MAINTENANCE_BR_UPDATE ||
      taskTypeId ===
        AppConstants.SERVER_CONSTANTS.TASK_TYPES.DUAL_MAINTENANCE_NEW_BR
    ) {
      return true;
    }
    return false;
  }
}
export default CommonUtil;
