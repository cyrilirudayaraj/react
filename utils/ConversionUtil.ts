import { EditorState, ContentState } from 'draft-js';
import htmlToDraft from 'html-to-draftjs';
import { forOwn } from 'lodash';
import moment from 'moment';
import { Shape } from '../types';
import StringUtil from './StringUtil';

class ConversionUtil {
  static convertMapToDropDownList(
    data: any,
    valueKey?: string,
    labelKey?: string
  ): string[] {
    const dropDownList = data?.map((val: any) => {
      return {
        value: val[valueKey ? valueKey : 'id'],
        text: val[labelKey ? labelKey : valueKey ? valueKey : 'name'],
      };
    });
    return dropDownList;
  }

  static convertMapToDropDownListInIdNameFormat(
    data: any,
    valueKey?: string,
    labelKey?: string
  ): string[] {
    const dropDownList = data?.map((val: any) => {
      const obj = {
        value: val[valueKey ? valueKey : 'id'],
        text:
          val[valueKey ? valueKey : 'id'] +
          ' - ' +
          val[labelKey ? labelKey : valueKey ? valueKey : 'name'],
      };
      return obj;
    });
    return dropDownList;
  }

  static convertDropDownListToValues(data: Shape[]): string[] {
    const values = data?.map((val: any) => {
      return val['value'];
    });
    return values;
  }

  static getListOfKeyValues(
    data: any[] | undefined,
    key: any
  ): string[] | undefined {
    const values = data?.map((val: any) => {
      return val[key];
    });
    return values;
  }

  static convertValuesToPayload(values: any): any {
    const payload: any = {};
    forOwn(values, function (value: any, key: string) {
      if (value instanceof Date) {
        payload[key] = value.toLocaleDateString('en-US');
      } else if (typeof value === 'boolean') {
        payload[key] = StringUtil.convertToYN(value);
      } else if (typeof value === 'string') {
        payload[key] = value ? value.trim() : value;
      } else {
        payload[key] = value;
      }
    });
    return payload;
  }

  static convertMinsToDhm(value: any): any {
    const duration = moment.duration(parseInt(value), 'minutes');
    let durationInMins = '';
    if (duration.days() > 0) {
      durationInMins += duration.days() + 'd ';
    }
    if (duration.hours() > 0) {
      durationInMins += duration.hours() + 'h ';
    }
    return durationInMins + duration.minutes() + 'm';
  }

  static convertMinsToHm(num: any): any {
    const hours: any = num / 60;
    const rhours: any = Math.floor(hours);
    const minutes: any = (hours - rhours) * 60;
    const rminutes: any = Math.round(minutes);
    return (
      (rhours > 0 ? rhours + 'h ' : ' ') + (rminutes > 0 ? rminutes + 'm' : '')
    );
  }

  static convertHtmlToPlainText(text: string): string {
    if (!text) {
      return text;
    }
    const editorState = EditorState.createWithContent(
      ContentState.createFromBlockArray(
        htmlToDraft(text).contentBlocks,
        htmlToDraft(text).entityMap
      )
    );
    return editorState.getCurrentContent().getPlainText();
  }

  static constructPayloadObj = (filters: any): any => {
    const ids: any = [];
    if (Array.isArray(filters)) {
      filters?.map((filter: any) => {
        if (typeof filter === 'object') {
          ids.push(filter.value);
        } else {
          ids.push(filter);
        }
      });
    }
    return ids;
  };

  static getPrettyJson(jsonString: string): string {
    try {
      const json = JSON.parse(jsonString);
      return JSON.stringify(json, null, 4);
    } catch (e) {
      return jsonString;
    }
  }
}

export default ConversionUtil;
