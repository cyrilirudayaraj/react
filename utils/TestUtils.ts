//TODO exclude this file during bundling
export function findByName(wrapper: any, name: string): any {
  const selector = "input[name='" + name + "']";
  return wrapper.find(selector).last();
}

export function findById(wrapper: any, id: string): any {
  const selector = '#' + id;
  return wrapper.find(selector).last();
}

export function changeValueById(wrapper: any, id: string, value: string): void {
  findById(wrapper, id).simulate('change', {
    target: { name: id, value: value },
  });
}

export function sleep(x: number): Promise<any> {
  return new Promise<void>((resolve) => {
    setTimeout(() => {
      resolve();
    }, x);
  });
}
