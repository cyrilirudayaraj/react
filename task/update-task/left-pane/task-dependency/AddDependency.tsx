import { Button, Form, FormField, Lightbox, Select } from '@athena/forge';
import { FieldArray, Formik, FormikProps } from 'formik';
import React, { Component } from 'react';
import * as Yup from 'yup';
import Labels from '../../../../../constants/Labels';
import Messages from '../../../../../constants/Messages';
import {
  DependencyDetails,
  TaskDependencyDetails,
  TaskDetail,
} from '../../../../../types';
import './TaskDependency.scss';
import { values, includes, find, remove } from 'lodash';
import DependencyReason from './DependencyReason';
import DependencySystem from './DependencySystem';
import AppConstants from '../../../../../constants/AppConstants';
import { fetchDependencyDetailsListByIds } from '../../../../../services/CommonService';
import { autoCompleteBlockedbyDependencies } from './TaskDependency';

export interface AddDependencyProps {
  show: boolean;
  onCancel?: any;
  onSave?: any;
  task?: TaskDetail;
}

const labels = Labels.TASK_DEPENDENCY;

const dependencyConditions =
  AppConstants.UI_CONSTANTS.TASK_DEPENDENCIES.DEPENDENCY_CONDITIONS;

const TASK_DEPENDENCY = AppConstants.UI_CONSTANTS.TASK_DEPENDENCY;

function validateByDependencyCondition(
  context: any,
  conditions: string[]
): boolean {
  //@ts-ignore
  const { dependencyCondition } = context.from[1].value;
  return includes(conditions, dependencyCondition);
}

export default class AddDependency extends Component<AddDependencyProps, any> {
  initialValues = {
    dependencyCondition: '',
    taskDependencies: [],
  };

  validationSchema = Yup.object().shape({
    dependencyCondition: Yup.string().required(Messages.MSG_REQUIRED),
    taskDependencies: Yup.array()
      .of(
        Yup.object().shape({
          // Deploy with/Test with related validations
          dependencySystemId: Yup.string()
            .nullable()
            .test('dependency-system-id', Messages.MSG_REQUIRED, function (
              value
            ) {
              let valid = true;
              if (
                validateByDependencyCondition(this, [
                  TASK_DEPENDENCY.DEPLOY_WITH,
                  TASK_DEPENDENCY.TEST_WITH,
                ])
              ) {
                valid = !!value;
              }
              return valid;
            }),
          dependencyId: Yup.string()
            .nullable()
            .test('dependency-id', Messages.MSG_REQUIRED, function (value) {
              let valid = true;
              if (
                validateByDependencyCondition(this, [
                  TASK_DEPENDENCY.DEPLOY_WITH,
                  TASK_DEPENDENCY.TEST_WITH,
                ])
              ) {
                valid = !!value;
              }
              return valid;
            })
            .test('dependency-id-invalid', Messages.INVALID, function (value) {
              let valid = true;
              if (
                validateByDependencyCondition(this, [
                  TASK_DEPENDENCY.DEPLOY_WITH,
                  TASK_DEPENDENCY.TEST_WITH,
                ])
              ) {
                valid = this.parent.dependencyName ? true : false;
              }
              return valid;
            })
            .test(
              'dependency-id-invalid-scope',
              Messages.TASK_DEPENDENCY_INVALID_SCOPE,
              function (value) {
                let valid = true;

                //@ts-ignore
                const { dependencyCondition } = this.from[1].value;
                const { TEST_WITH, ATLAS_ID } = TASK_DEPENDENCY;
                if (
                  dependencyCondition == TEST_WITH &&
                  this.parent.dependencySystemId == ATLAS_ID
                ) {
                  valid =
                    this.parent.activeWorkflowStepId <=
                    AppConstants.SERVER_CONSTANTS.WORKFLOW_STEPS
                      .TEST_CHANGES_STEP
                      ? true
                      : false;
                }
                return valid;
              }
            )
            .test({
              name: 'dependency-id-duplicate',
              message: Messages.DUPLICATE_ENTRY,
              test: ((props) => {
                return function (value: any) {
                  let valid = true;
                  const context = {
                    //@ts-ignore
                    ...this,
                  };
                  if (
                    validateByDependencyCondition(context, [
                      TASK_DEPENDENCY.DEPLOY_WITH,
                      TASK_DEPENDENCY.TEST_WITH,
                    ])
                  ) {
                    const {
                      dependencyCondition,
                      taskDependencies: newDependencies,
                    } = context.from[1].value;
                    const { parent } = context;

                    // Check any duplicate dependency id found in newly created dependencies
                    valid = !newDependencies.some(
                      (newDependency: TaskDependencyDetails) => {
                        return (
                          newDependency != parent &&
                          parent.dependencyId == newDependency.dependencyId &&
                          parent.dependencySystemId ==
                            newDependency.dependencySystemId
                        );
                      }
                    );

                    // Check any duplicate dependency id found in existing dependencies
                    const taskDependencies = props.task?.taskDependencies;
                    if (taskDependencies && valid) {
                      const { dependencySystemId } = context.parent;

                      valid = !taskDependencies.some(
                        (taskDependency, index) => {
                          return (
                            taskDependency.dependencyId == value &&
                            taskDependency.dependencySystemId ==
                              dependencySystemId &&
                            taskDependency.dependencyCondition ==
                              dependencyCondition
                          );
                        }
                      );
                    }
                  }
                  return valid;
                };
              })(this.props),
            }),
          // Blocked by related validations
          dependencyName: Yup.string().test(
            'dependency-name',
            Messages.MSG_REQUIRED,
            function (value) {
              let valid = true;
              if (
                validateByDependencyCondition(this, [
                  TASK_DEPENDENCY.BLOCKED_BY,
                ])
              ) {
                valid = !!value;
              }
              return valid;
            }
          ),
          dependencyDate: Yup.date()
            .nullable()
            .when('dependencyName', {
              is:
                AppConstants.UI_CONSTANTS.TASK_DEPENDENCY
                  .SPECIFIC_DEPLOYMENT_DATE,
              then: Yup.date().required(Messages.MSG_REQUIRED),
              otherwise: Yup.date(),
            }),
        })
      )
      .min(1),
  });

  createNewTaskDependency = (): TaskDependencyDetails => {
    const taskDependency: TaskDependencyDetails = {
      dependencyCondition: '',
      dependencyDate: '',
      dependencyId: '',
      dependencyName: '',
      dependencySystemId: '',
      dependencySystemName: '',
      deploymentDate: '',
      id: '',
      notes: '',
      ordering: '',
      statusName: '',
      taskId: '',
      version: '',
      completedYn: '',
      description: '',
    };
    return taskDependency;
  };

  onAdd = (index: number, arrayHelper: any) => {
    arrayHelper.push(this.createNewTaskDependency());
  };

  onDelete = (index: number, arrayHelper: any) => {
    arrayHelper.remove(index);
  };

  renderDependencyByCondition(
    formik: FormikProps<any>,
    dependencyProps: any
  ): JSX.Element | undefined {
    let field: any;
    switch (formik.values.dependencyCondition) {
      case TASK_DEPENDENCY.DEPLOY_WITH:
      case TASK_DEPENDENCY.TEST_WITH: {
        field = <DependencySystem {...dependencyProps} />;
        break;
      }
      case TASK_DEPENDENCY.BLOCKED_BY: {
        field = <DependencyReason {...dependencyProps} />;
        break;
      }
    }
    return field;
  }

  renderDepedencyTaskTemplate = (formik: FormikProps<any>): any => {
    let content = null;

    const taskDependencies: TaskDependencyDetails[] =
      formik.values.taskDependencies;

    content = (
      <FieldArray
        name="taskDependencies"
        render={(arrayHelpers) => {
          if (taskDependencies.length == 0) {
            this.onAdd(-1, arrayHelpers);
          }

          return (
            <div>
              {taskDependencies.map(
                (obj: TaskDependencyDetails, index: number) => {
                  const dependencyProps: any = {
                    index: index,
                    formik: formik,
                    task: this.props.task,
                    onAdd: (index: number) => this.onAdd(index, arrayHelpers),
                    onDelete: (index: number) =>
                      this.onDelete(index, arrayHelpers),
                  };

                  return this.renderDependencyByCondition(
                    formik,
                    dependencyProps
                  );
                }
              )}
            </div>
          );
        }}
      />
    );

    return content;
  };

  beforeFormSubmit(formik: FormikProps<any>) {
    const { dependencyCondition, taskDependencies } = formik.values;
    if (
      dependencyCondition == TASK_DEPENDENCY.DEPLOY_WITH ||
      dependencyCondition == TASK_DEPENDENCY.TEST_WITH
    ) {
      const payload = taskDependencies.map((o: TaskDependencyDetails) => {
        const { dependencyId, dependencySystemId } = o;
        return { dependencyId, dependencySystemId };
      });

      fetchDependencyDetailsListByIds(payload).then(
        (result: DependencyDetails[]) => {
          const updatedDeps: TaskDependencyDetails[] = taskDependencies.map(
            (o: TaskDependencyDetails) => {
              const dependencyDetail: DependencyDetails | undefined = find(
                result,
                {
                  dependencyId: o.dependencyId,
                  dependencySystemId: o.dependencySystemId,
                }
              );
              const data = {
                ...o,
                deploymentDate: dependencyDetail?.deploymentDate,
                dependencyName: dependencyDetail?.dependencyName,
                statusName: dependencyDetail?.statusName,
                activeWorkflowStepId: dependencyDetail?.activeWorkflowStepId,
              };
              return data;
            }
          );
          formik.setFieldValue('taskDependencies', updatedDeps, false);
          formik.handleSubmit();
        }
      );
    } else {
      const deploymentDate = this.props.task?.deploymentDate;
      if (deploymentDate) {
        formik.setFieldValue(
          'taskDependencies',
          autoCompleteBlockedbyDependencies(
            taskDependencies,
            deploymentDate,
            dependencyCondition
          ),
          false
        );
      }
      formik.handleSubmit();
    }
  }

  renderForm(formik: FormikProps<any>): JSX.Element {
    const { task } = this.props;
    const { TEST_WITH } = AppConstants.UI_CONSTANTS.TASK_DEPENDENCY;
    const dependencyConditionValues = values(dependencyConditions);
    if (
      task?.activeWorkflowStepId &&
      task?.activeWorkflowStepId >
        AppConstants.SERVER_CONSTANTS.WORKFLOW_STEPS.TEST_CHANGES_STEP
    )
      remove(dependencyConditionValues, function (object: any) {
        return object.value === TEST_WITH;
      });
    return (
      <Form
        className="fe_u_margin--medium"
        layout="compact"
        labelAlwaysAbove
        includeSubmitButton={false}
        autoComplete="off"
        name="form"
      >
        <div className="row">
          <FormField
            className="row2"
            id="dependencyCondition"
            required={true}
            inputAs={Select}
            labelText={labels.TYPE}
            options={dependencyConditionValues}
            {...formik.getFieldProps('dependencyCondition')}
            onChange={(event: any) => {
              formik.setValues(
                {
                  ...formik.values,
                  dependencyCondition: event.target.value,
                  taskDependencies: [],
                },
                false
              );
            }}
            error={formik.errors.dependencyCondition?.toString()}
          />
        </div>

        {this.renderDepedencyTaskTemplate(formik)}
      </Form>
    );
  }

  render(): JSX.Element {
    return (
      <Formik
        initialValues={this.initialValues}
        validationSchema={this.validationSchema}
        onSubmit={this.props.onSave}
        validateOnMount={false}
        validateOnChange={false}
        validateOnBlur={false}
      >
        {(formik) => {
          return (
            <Lightbox
              show={this.props.show}
              hideDividers
              headerText={labels.ADD_DEPENDENCIES}
              disableClose
              width="large"
              className="my-custom-lightbox add-dependency"
            >
              <Button
                variant="tertiary"
                icon="Close"
                onClick={this.props.onCancel}
                className="my-close-button"
              />
              {this.renderForm(formik)}
              <div className="fe_c_lightbox__footer">
                <Button
                  text={labels.CANCEL}
                  variant="secondary"
                  className="fe_u_margin--right-small"
                  onClick={this.props.onCancel}
                />
                <Button
                  text={labels.SAVE}
                  onClick={() => this.beforeFormSubmit(formik)}
                />
              </div>
            </Lightbox>
          );
        }}
      </Formik>
    );
  }
}
