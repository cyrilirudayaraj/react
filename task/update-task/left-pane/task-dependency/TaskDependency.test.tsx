import { configure, mount } from 'enzyme';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import TaskDependency from './TaskDependency';
import { Formik, FormikValues } from 'formik';
import { sleep } from '../../../../../utils/TestUtils';
import AddDependency from './AddDependency';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { TaskDetail, TaskStep } from '../../../../../types/index';

jest.mock('../../../../../services/CommonService');
configure({ adapter: new Adapter() });
let store: any;
let taskDetails: TaskDetail;
beforeEach(() => {
  taskDetails = {} as TaskDetail;
  taskDetails.activeTaskStepId = '1';
  const taskStep: TaskStep = {} as TaskStep;
  taskDetails.taskSteps = [] as TaskStep[];
  taskDetails.taskSteps[0] = taskStep;
  taskDetails.taskSteps[0].name = 'Analyze Task';
  taskDetails.taskSteps[0].workflowStepId = '1';
  store = configureStore([])({
    task: {
      taskDetails: taskDetails,
    },
    userPermission: {
      userPermissions: ['task.update'],
    },
  });
});
describe('TaskDependency', () => {
  const props: FormikValues = {
    values: {
      index: 0,
      taskDependencies: [
        {
          deploymentDate: null,
          dependencyName: 'provi"der" number, missing, test 2',
          dependencyId: '30',
          taskId: '32',
          dependencyDate: '12/10/2020',
          version: '3',
          status: 'Pending Deployment',
          dependencySystemName: 'Atlas',
          description: null,
          ordering: null,
          dependencySystemId: '3',
          notes: 'test messages',
          type: 'deploy with',
          id: '1',
          completedYn: null,
        },
        {
          deploymentDate: null,
          dependencyName: 'FCN - Task Dependencies (View)',
          dependencyId: 'COLRDD-19169',
          taskId: '32',
          dependencyDate: null,
          version: '3',
          status: 'Code Review',
          dependencySystemName: 'Jira',
          description: null,
          ordering: null,
          dependencySystemId: '2',
          notes: null,
          type: 'deploy with',
          id: '2',
          completedYn: null,
        },
        {
          deploymentDate: '11/17/2020',
          dependencyName: 'provider number missing test 3',
          dependencyId: '33',
          taskId: '32',
          dependencyDate: null,
          version: '10',
          status: 'Deployed',
          dependencySystemName: 'Atlas',
          description: null,
          ordering: null,
          dependencySystemId: '3',
          notes: 'test message',
          type: 'test with',
          id: '3',
          completedYn: null,
        },
        {
          deploymentDate: null,
          dependencyName: 'UI changes',
          dependencyId: 'COLRDD-19488',
          taskId: '32',
          dependencyDate: null,
          version: '2',
          status: 'New',
          dependencySystemName: 'Jira',
          description: null,
          ordering: null,
          dependencySystemId: '2',
          notes: 'blocked by test message',
          type: 'blocked by',
          id: '4',
          completedYn: null,
        },
        {
          deploymentDate: null,
          dependencyName:
            'Local GA - Columbus Regional (FPC/OPC) [226] - Insurance Default - Override 195 for package 17045',
          dependencyId: '12356',
          taskId: '32',
          dependencyDate: null,
          version: '2',
          status: 'INPRODUCTION',
          dependencySystemName: 'Rule Tracker',
          description: null,
          ordering: null,
          dependencySystemId: '1',
          notes: null,
          type: 'deploy with',
          id: '5',
          completedYn: null,
        },
        {
          deploymentDate: '11/17/2020',
          dependencyName: 'provider number missing test 3',
          dependencyId: '33',
          taskId: '32',
          dependencyDate: null,
          version: '2',
          status: 'In Progres',
          dependencySystemName: 'Atlas',
          description: null,
          ordering: null,
          dependencySystemId: '3',
          notes: 'Specific deployment date test',
          type: 'blocked by',
          id: '6',
          completedYn: null,
        },
      ],
    },
    touched: {},
    errors: {},
    handleChange: (a: any) => a,
    handleBlur: (a: any) => a,
    handleSubmit: (a: any) => a,
    getFormFieldProps: jest.fn((fieldName: string, formik: any) => {
      return { value: 'test' };
    }),
    getFieldProps: jest.fn(() => {
      return { value: 'tets' };
    }),
  };

  let wrapper: any;
  let outerProps: any;

  beforeEach(() => {
    const { taskDependencies } = props.values;

    wrapper = mount(
      <Formik
        initialValues={{
          taskDependencies,
        }}
        onSubmit={jest.fn()}
        validate={jest.fn()}
      >
        {(formik) => {
          outerProps = {
            formik,
          };
          return (
            <Provider store={store}>
              <TaskDependency {...outerProps} />
            </Provider>
          );
        }}
      </Formik>
    );
  });

  it('should render as expected', () => {
    //@ts-ignore
    wrapper = mount(
      <Provider store={store}>
        <TaskDependency formik={props}></TaskDependency>
      </Provider>
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should open create dependency dialog when add clicked', async () => {
    wrapper.find('.root-add-button').last().simulate('click');
    await sleep(10);

    expect(wrapper.find(AddDependency).length).toEqual(1);
  });

  it('should close create dependency dialog when cancel clicked', async () => {
    wrapper.find('.root-add-button').last().simulate('click');
    await sleep(10);

    expect(wrapper.find(AddDependency).length).toEqual(1);
    wrapper.find('.fe_c_button--secondary').first().simulate('click');
    await sleep(10);
    expect(wrapper.find(AddDependency).length).toEqual(0);
  });
});
