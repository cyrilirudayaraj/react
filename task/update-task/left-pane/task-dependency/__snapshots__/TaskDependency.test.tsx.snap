// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDependency should render as expected 1`] = `
"<Provider store={{...}}>
  <Connect(TaskDependency) formik={{...}}>
    <TaskDependency formik={{...}} task={{...}} refreshTaskList={[Function (anonymous)]}>
      <div className=\\"task-dependency\\">
        <div className=\\"task-dependency-details\\">
          <FormikConnect(FieldArrayInner) name=\\"taskDependencies\\" render={[Function: render]}>
            <FieldArrayInner name=\\"taskDependencies\\" render={[Function: render]} formik={[undefined]} validateOnChange={true}>
              <div>
                <div>
                  <div>
                    <div className=\\"add-dependency\\">
                      <Connect(WFEnableForPermission) permission=\\"task.dependency-details.update\\">
                        <WFEnableForPermission permission=\\"task.dependency-details.update\\" userPermissions={{...}} dispatch={[Function: dispatch]}>
                          <ForwardRef className=\\"root-add-button\\" text=\\"Add\\" icon=\\"Add\\" variant=\\"tertiary\\" onClick={[Function: onClick]} disabled={false} fullWidth={false} iconSet={{...}} size=\\"medium\\" showLoadingSpinner={false} type=\\"button\\" useLink={false}>
                            <button className=\\"fe_c_button fe_c_button--tertiary fe_c_button--medium root-add-button\\" aria-disabled={false} aria-label={[undefined]} onClick={[Function: handleClick]} type=\\"button\\">
                              <div className=\\"fe_c_button__content\\">
                                <Add title={{...}} disabled={false} className=\\"fe_c_button__icon fe_c_button__icon--with-text fe_c_button__icon--tertiary\\">
                                  <IconBase viewBox=\\"0 0 25 25\\" title={{...}} disabled={false} className=\\"fe_c_button__icon fe_c_button__icon--with-text fe_c_button__icon--tertiary\\">
                                    <svg xmlns=\\"http://www.w3.org/2000/svg\\" className=\\"fe_c_icon fe_c_button__icon fe_c_button__icon--with-text fe_c_button__icon--tertiary\\" viewBox=\\"0 0 25 25\\" style={{...}}>
                                      <g>
                                        <g className=\\"fe_c_icon__background\\">
                                          <rect width=\\"25\\" height=\\"25\\" />
                                        </g>
                                        <g className=\\"fe_c_icon__icon fe_c_icon__icon--primary\\">
                                          <g>
                                            <polygon points=\\"13.3 7.8 11.8 7.8 11.8 11.8 7.8 11.8 7.8 13.3 11.8 13.3 11.8 17.2 13.3 17.2 13.3 13.3 17.2 13.3 17.2 11.8 13.3 11.8 13.3 7.8\\" />
                                            <path d=\\"M20,5a11,11,0,0,0-7.5-3.1A11,11,0,0,0,5,5,10.7,10.7,0,0,0,5,20a11,11,0,0,0,7.5,3.1A11,11,0,0,0,20,20,10.7,10.7,0,0,0,20,5ZM18.9,18.9a9.1,9.1,0,0,1-6.4,2.7,9.1,9.1,0,0,1,0-18.2,9.1,9.1,0,0,1,6.4,15.5Z\\" />
                                          </g>
                                        </g>
                                      </g>
                                    </svg>
                                  </IconBase>
                                </Add>
                                <span className=\\"fe_c_button__text fe_c_button__text--medium fe_c_button__text--tertiary\\">
                                  Add
                                </span>
                              </div>
                            </button>
                          </ForwardRef>
                        </WFEnableForPermission>
                      </Connect(WFEnableForPermission)>
                    </div>
                  </div>
                  <div>
                    <div className=\\"deploy-with\\">
                      <b />
                    </div>
                    <div className=\\"test-with\\">
                      <b />
                    </div>
                    <div className=\\"blocked-by\\">
                      <b />
                    </div>
                  </div>
                </div>
              </div>
            </FieldArrayInner>
          </FormikConnect(FieldArrayInner)>
        </div>
      </div>
    </TaskDependency>
  </Connect(TaskDependency)>
</Provider>"
`;
